mlx_lm/awq.py

```diff
@@ -0,0 +1,535 @@
+    elif isinstance(prev_op, (nn.LayerNorm, nn.RMSNorm)):
+        prev_op.weight = prev_op.weight / scales
+        if hasattr(prev_op, "bias"):
+            prev_op.bias = prev_op.bias / scales
+        for layer in layers:
+            layer.weight = layer.weight * scales
+    elif prev_op.__class__.__name__ == "RMSNorm":  # For gemma models
+        prev_op.weight = (1.0 + prev_op.weight) / scales - 1.0
+        for layer in layers:
+            layer.weight = layer.weight * scales
+    else:
+        raise NotImplementedError(f"Could not apply scale to prev_op: {prev_op}")
+
+    for layer in layers:
+        layer.input_feat = layer.input_feat / scales
+
+
+@dataclass
+class LayerConfig:
+    prev_op: nn.Module
+    layers: list[nn.Module]
+    block: nn.Module | None = None
+    layer_kwargs: dict | None = None
+
+
+def scale_block(
+    block: nn.Module, mask: mx.array, quantize_func: Callable, n_grid: int = 20
+):
+    # Layers which have the same inputs (within an elementwise multiplication)
+    # can be scaled together.
+    mlp_norm = getattr(
+        block, "pre_feedforward_layernorm", block.post_attention_layernorm
+    )
+    config = [
+        LayerConfig(
+            block=block.self_attn,
+            prev_op=block.input_layernorm,
+            layers=[
+                block.self_attn.q_proj,
+                block.self_attn.k_proj,
+                block.self_attn.v_proj,
+            ],
+            layer_kwargs={"mask": mask},
+        )
+    ]
+    # Llama 4 MOE layer
+    if hasattr(block, "feed_forward") and block.is_moe_layer:
+        config += [
+            LayerConfig(
+                block=block.feed_forward,
+                prev_op=block.post_attention_layernorm,
+                layers=[
+                    block.feed_forward.shared_expert.gate_proj,
+                    block.feed_forward.shared_expert.up_proj,
+                    block.feed_forward.experts.gate_proj,
+                    block.feed_forward.experts.up_proj,
+                    block.feed_forward.router,
+                ],
+            ),
+            LayerConfig(
+                prev_op=block.feed_forward.experts.up_proj,
+                layers=[
+                    block.feed_forward.experts.down_proj,
+                ],
+            ),
+            LayerConfig(
+                prev_op=block.feed_forward.shared_expert.up_proj,
+                layers=[
+                    block.feed_forward.shared_expert.down_proj,
+                ],
+            ),
+        ]
+    else:
+        config += [
+            LayerConfig(
+                block=block.mlp,
+                prev_op=mlp_norm,
+                layers=[
+                    block.mlp.gate_proj,
+                    block.mlp.up_proj,
+                ],
+            ),
+            LayerConfig(
+                prev_op=block.mlp.up_proj,
+                layers=[
+                    block.mlp.down_proj,
+                ],
+            ),
```

Comment:
It's a little confusing to me that we find an optimal scale for the `gate_proj` and `up_proj` but then we fuse the scale found for the `down_proj` into the `up_proj` which could revert the improvement for the `up_proj`. Maybe we should search and apply the `down_proj` before we search for the `gate_proj` and `up_proj`.